notebooks/development/test_ghani_preprocessing.ipynb

```diff
@@ -0,0 +1,1199 @@
+     ]
+    }
+   ],
+   "source": [
+    "from birdset.modules.models.birdnet import BirdNetModel\n",
+    "import torch.nn as nn\n",
+    "\n",
+    "num_classes = 31 #! Don't forget to change this\n",
+    "sampling_rate = 48_000 \n",
+    "window_length = 3\n",
+    "input_size = 1024\n",
+    "\n",
+    "perch_network = BirdNetModel(num_classes=num_classes, model_path='../../checkpoints/birdnet/BirdNET_GLOBAL_6K_V2.4_Model', train_classifier=False)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 3. Batch and Preprocess the dataset "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### 3.1. k-sample the dataset"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Number of samples in the training set: 664\n",
+      "Number of samples in the validation set: 221\n",
+      "Number of samples in the testing set: 221\n"
+     ]
+    }
+   ],
+   "source": [
+    "print(\"Number of samples in the training set:\", len(dataset['train']))\n",
+    "print(\"Number of samples in the validation set:\", len(dataset['valid']))\n",
+    "print(\"Number of samples in the testing set:\", len(dataset['test']))"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Number of samples in the training set: 480\n",
+      "Number of samples in the validation set: 313\n",
+      "Number of samples in the testing set: 313\n"
+     ]
+    }
+   ],
+   "source": [
+    "import random\n",
+    "from collections import defaultdict\n",
+    "from datasets import concatenate_datasets, DatasetDict, Dataset\n",
+    "\n",
+    "# Define the number of samples per class\n",
+    "samples_per_class = 30\n",
+    "\n",
+    "# Merge the train, valid, and test splits\n",
+    "merged_data = concatenate_datasets([dataset['train'], dataset['valid'], dataset['test']])\n",
+    "merged_data = merged_data.shuffle()\n",
+    "\n",
+    "# Create a dictionary to store the selected samples per class\n",
+    "selected_samples = defaultdict(list)\n",
+    "rest_samples = []\n",
+    "# Iterate over the merged data and select the desired number of samples per class\n",
+    "for sample in merged_data:\n",
+    "    label = sample['labels']\n",
+    "    if len(selected_samples[label]) < samples_per_class:\n",
+    "        selected_samples[label].append(sample)\n",
+    "    else:\n",
+    "        rest_samples.append(sample)    \n",
+    "\n",
+    "# Flatten the selected samples into a single list\n",
+    "selected_samples = [sample for samples in selected_samples.values() for sample in samples]\n",
+    "\n",
+    "# Split the selected samples into training, validation, and testing sets\n",
+    "test_ratio = 0.5\n",
```

Comment:
@XgamerTV Validation is not really common in such a few shot setting and since the dataset is quite small I would suggest to keep all remaining samples in the test set. Isnt that what Ghani is also doing?

Reply:
@raphaelschwinger Yeah I totally agree maybe the smartest thing would be to make validation and test set the same so that you can still see the results after every epoch or what do you have in mind🤔

Reply:
@raphaelschwinger Ok! A general question: How does the validation set impact the training I thought it was just for manual hyperparameter tuning. Training till convergence seems good for the final results and is also what Ghani do.

Reply:
@XgamerTV As you describe, for manual and automatic hyperparameter tuning. But using information about the test set to find the best hyperparameter is not possible in practice and therefore should be strongly avoided.

Reply:
Ok thx I understand why it doesn't make sense from a scientific pov 👍